builder/resubmit_utils.go

```diff
@@ -9,13 +9,28 @@ import (
 )
 
 // runResubmitLoop checks for update signal and calls submit respecting provided rate limiter and context
-func runResubmitLoop(ctx context.Context, limiter *rate.Limiter, updateSignal chan struct{}, submit func()) {
+func runResubmitLoop(ctx context.Context, limiter *rate.Limiter, updateSignal chan struct{}, submit func(), submitTime *time.Time) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-updateSignal:
-			res := limiter.Reserve()
+			var res *rate.Reservation
+			if submitTime == nil {
```

Comment:
Can `submitTime` be nil? If not, maybe it would be better to assert it's not nil and require that it be passed in

Reply:
good point, based on your other comment i've refactored this, please feel free to unresolve if there are still concerns